vllm/model_executor/sampling_metadata.py

```diff
@@ -6,54 +6,269 @@
+        return len(self.sample_indices) > 0
+
+    def __post_init__(self):
+        if len(self.prefill_indices) > 0:
+            assert self.sampling_params.prompt_logprobs is not None
+
+
 class SamplingMetadata:
     """Metadata for input sequences. Used in sampler.
 
+    The usage is as follow;
+    ```
+    logits = execute_model(...)
+    # prune logits
+    logits[sampling_metadata.selected_token_indices]
+    sample(logits)
+
+    def sample(logits):
+        # Use categorized_sample_indices....
+    ```
+
+    categorized_sample_indices is index after the pruning.
+
     Args:
-        seq_groups: List of (seq_ids, sampling_params).
-        seq_data: Seq_id -> SequenceData.
-        prompt_lens: Lengths of prompts.
-        selected_token_indices: Token indices selected for sampling.
+        seq_groups: List of batched sequence groups.
+        selected_token_indices: (num_query_tokens_to_logprob), Query token
+            indices to process for sampling or logprob calculation.
         categorized_sample_indices: SamplingType -> token indices to sample.
-        generators: List of torch.Generators to use for seeded sampling
+            Each token indices is 2D tensor of (num_indices, num_indices) where
+            the first item means the sample index within the returned logit,
+            and the second item means the sample index within selected tokens.
+            For example, if the returned logit is [1, 2, 3], and we select
+            [1, 2] for sampling, the pruned logit will be [2, 3]. The first
+            tuple is [1, 2] (sampled index within original logit), and the
+            second tuple is [0, 1] (sampled index within pruned logit).
         perform_sampling: Whether to perform sampling. This option is used to
             make the sampling only happens in the driver worker, and disable
-            sampling in other worker processes.
+            sampling in other worker processes. Setting this True is equivalent
+            to choose no indices for categorized_sample_indices.
     """
 
     def __init__(
         self,
-        seq_groups: Optional[List[Tuple[List[int], SamplingParams]]],
-        seq_data: Optional[Dict[int, SequenceData]],
-        prompt_lens: Optional[List[int]],
+        seq_groups: List[SequenceGroupToSample],
         selected_token_indices: torch.Tensor,
         categorized_sample_indices: Optional[Dict[SamplingType, torch.Tensor]],
-        generators: Optional[List[torch.Generator]] = None,
+        num_prompts: int,
         perform_sampling: bool = True,
     ) -> None:
         self.seq_groups = seq_groups
-        self.seq_data = seq_data
-        self.prompt_lens = prompt_lens
         self.selected_token_indices = selected_token_indices
         self.categorized_sample_indices = categorized_sample_indices
-        self.generators = generators
         self.perform_sampling = perform_sampling
+        self.num_prompts = num_prompts
+
+    @staticmethod
+    def prepare(
```

Comment:
Main delta
1. do_sample is introduced
2. it creates `SequenceGroupToSample`
3. Prepare prefill_indices and sample_indices for each seq group